Bug 2014352: Could not filter out machine by using node name on machines page

kdoberst opened the pull request:
This address [Bug 2014352](https://bugzilla.redhat.com/show_bug.cgi?id=2014352)

In version 4.8, on the machine list when using the "name" filter, the filter would accept matches from both the "name" and "node" columns of the table.  In version 4.9, this changed so only matches from the "name" column were considered.

This PR brings the 4.8 functionality back.

**Filtering machine by name**
<img width="500" alt="Screen Shot 2021-11-10 at 1 03 38 PM" src="https://user-images.githubusercontent.com/82059948/141491933-1771153d-7159-46a6-aff6-0fe29f354edd.png">


**Filtering machine by node**
<img width="500" alt="Screen Shot 2021-11-10 at 1 04 05 PM" src="https://user-images.githubusercontent.com/82059948/141177290-3a0651f0-9eb2-42b3-8cfe-ee61f28bc122.png">

rawagner commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/console-types.ts:
```diff
@@ -396,6 +396,7 @@ export type ListPageFilterProps<D = any> = {
   columnLayout?: ColumnLayout;
   onFilterChange: OnFilterChange;
   hideColumnManagement?: boolean;
+  textFilter?: string;
```
This is not ok, `textFilter` loosely references a function from `table-filters.ts`. How do you expect it to work for dynamic plugins that wont contribute function to `table-filters` ? We need to get away from those loose references and not re-introduce them.

kdoberst replied:
Removing the textFilters prop and passing custom filters via the rowFilters param for the useListPageFilter hook

rawagner commented on frontend/public/components/machine.tsx:
```diff
@@ -298,6 +298,7 @@ export const MachinePage: React.FC<MachinePageProps> = ({
           hideNameLabelFilters={hideNameLabelFilters}
           hideLabelFilter={hideLabelFilter}
           hideColumnManagement={hideColumnManagement}
+          textFilter="machine"
```
as I mentioned above, this is not a correct approach.

We need to allow user to modify the default `name` filter via `useListPageFilter`.
`useListPageFilter` already accepts custom `rowFilters`. The custom `name` filter logic could be passed as a part of `rowFilters` arg or we can add a new arg.

kdoberst replied:
I'm not sure why, but I missed that you could pass additional filters to the useListPageFilter hook.  Using the rowFilters param to pass in a custom filter and removing the textFilter props.

rawagner commented on frontend/public/components/machine.tsx:
```diff
@@ -281,7 +282,17 @@ export const MachinePage: React.FC<MachinePageProps> = ({
   });
 
-  const [data, filteredData, onFilterChange] = useListPageFilter(machines);
+  const filters: RowFilter[] = [
```
looks much better :)

Can you update `RowFilter` type so it can contain only `type` and `filter` ?

currently we have 
```js
export type RowFilter<R = any> = RowMatchFilter<R> | RowReducerFilter<R>;
```

so we could add something like (maybe with better naming)
```js
type SimpleRowFilter<R> = {
  type: string;
  filter: (input: FilterValue, obj: R) => boolean;
}

export type RowFilter<R = any> = RowMatchFilter<R> | RowReducerFilter<R> | SimpleRowFilter<R>;
```

kdoberst commented:
@rawagner @spadgett 
The core of this issue is that instead of the "name" filter, the "name" filter should actually use the "machine" filter.  So I pass a custom filter (that just points the name filter to the machine filter). 

The problem is that with the types as written, I needed to pass in items that weren't actually needed (filterGroupName, Items, defautSelected).   As Rastislav pointed out, this isn't ideal.  So I created a "RowSimpleFilter" type to account for this.   In order to prevent typescript errors, I needed to add the definition to both the `filter-toolbar.tsx` and `console-types.ts`. It appears the code is pulling types from both areas.  In addition, I needed to add optional items to the RowSimpleRow definition in the `filter-toolbar.tsx` file.  Since this is a bug fix, I'm a little weary to making changes that have the potential to cause issues across the application.

So after all that, here are my questions:
1.  Did I miss something obvious (I'm fairly new to Typescript)?
2. Anything else I need to consider considering I'm changing the dynamic plugin sdk?
3.  This does seem straightforward or should we take the "RowSimpleFilter" definition out and add the extra params back into the "filters" array object (machine.tsx line ~285)?

rawagner commented on frontend/public/components/filter-toolbar.tsx:
```diff
@@ -421,6 +421,14 @@ type RowFilterBase<R> = {
   defaultSelected?: string[];
 };
 
+export type RowSimpleFilter<R = any> = {
```
why is this different than `RowSimpleFilter` from `console-types` ?

kdoberst replied:
There were all sorts of type errors when I used the definition from console-types in filter-toolbar.  Hopefully having all areas look at console-types, will resolve the errors I was seeing.

rawagner commented:
>
>     1. Did I miss something obvious (I'm fairly new to Typescript)?

I dont think so

> 
>     2. Anything else I need to consider considering I'm changing the dynamic plugin sdk?

Since you're only adding to the SDK, then we are fine.

> 
>     3. This does seem straightforward or should we take the "RowSimpleFilter" definition out and add the extra params back into the "filters" array object (machine.tsx line ~285)?

Actually RowSimpleFilter is there only for `name` and `label` filtering logic. So we could rename it to `RowNameLabelFilter` to make it more obvious. The `type` param can be narrowed down to `'name' | 'labels'` instead of `string`. So like this 
```js
export type RowSimpleFilter<R = any> = {
  type: 'name' | 'labels';
  filter: (input: FilterValue, obj: R) => boolean;
};
```

kdoberst commented on frontend/public/components/machine.tsx:
```diff
@@ -281,7 +282,14 @@ export const MachinePage: React.FC<MachinePageProps> = ({
   });
 
-  const [data, filteredData, onFilterChange] = useListPageFilter(machines);
+  const filters: RowSimpleFilter[] = [
+    {
+      type: 'name',
+      filter: tableFilters.machine,
+    },
+  ];
+
```
Good point - just happy to get it to work ;-). I'll move it.

kdoberst commented:
@rawagner I made changes and to hopefully help make things clear, I have two commits.  

For the first commit, I forced changes that rename the filter to `RowNameLabelFilter `.

For the second commit, I removed references to `RowMatchFilter`, `RowReducerFilter`, `RowFilter` and `RowNameLabelFilter` from `filter-toolbar.tsx` file.  So all files in the application should be getting the type definition from `console-types.ts`

In order to prevent type errors in `filter-toolbar.tsx` I added `filterGroupName`, `items`, and `defaultSelected` as optional on the `RowNameLabelFilter` (see console-types.ts line ~353).  Knowing that these optional items may be undefined, I added some checks in `filter-toolbar.tsx`.  To be specific, the following lines in `filter-toolbar.tsx` were throwing type errors without these optional items in the definition: 120, 137, 176.

If we want a single commit - I can rebase them together, just thought it might be easier to review if they were separate.  I can also point references to the above items in the code.  Thought giving a single summary might be easiser?

kdoberst commented:
/retest

rawagner commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/console-types.ts:
```diff
@@ -346,7 +346,15 @@ type RowFilterBase<R> = {
+  defaultSelected?: string[];
+};
+
+export type RowNameLabelFilter<R = any> = {
```
this type should have only `type` and `filter` props. Other have no meaning

kdoberst replied:
The problem is that when I added RowNameFilter to the RowFilter definition, there is code that depends on rilterGroupName and/or items and/or defaultSelected as part of the type definition.  There is code that is depending on items of RowFilter to have these items as part of the type definition.  By adding RowNameLabelFilter as an option of RowFilter, these options are not guaranteed to be part of the type definition and cause type errors.

See `/frontend/public/components/filter-toolbar.tsx`[ line 173](https://github.com/openshift/console/blob/master/frontend/public/components/filter-toolbar.tsx#L173) for an example.

kdoberst commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/console-types.ts:
```diff
@@ -346,7 +346,15 @@ type RowFilterBase<R> = {
   filterGroupName: string;
   type: string;
   items: RowFilterItem[];
+  filter?: (input: FilterValue, obj: R) => boolean;
```
This is from when I removed the RowFilterBase from the filter-toolbar.tsx file (and had all references to the filter-toolbar.tsx point here).

The old definition had filter as optional and there are numerous variables where filter is missing across the application.  A rough count is about 20 pages.  
See `/frontend/packages/console-app/src/components/nodes/NodesPage.tsx` line 498 as an example.

kdoberst commented:
@rawagner I've added some comments on why I needed to modify the type definitions.  I will be the first one to admit that the changes were far from ideal.  The more I dig into this and as I try to address the type definitions, I'm uncovering more and more issues.  Sorting it out, I feel, will take more changes across the application and more than I would like for what, in the end, should be a simple bug fix.  TypeScript is doing its job and letting us know that there is something deeper going on.

I'm proposing that we move the unifying the RowFilter definitions (remove the double definition), sorting out the filter type definitions,  and include a new "RowNameLabelFilter" to a tech-debt story.  To fix the issue associated with this bug, I'm proposing a simpler fix where, for right now, we put in a single `@ts-ignore`.  Hopefully, this ignore can be removed with the tech-debt story.  Here is what I was thinking: [simple change to machine filtering](https://github.com/openshift/console/compare/master...kdoberst:Bug-2014352-simple-fix-to-could-not-filter-out-by-using-nade-name?expand=1)

Thoughts?

kdoberst commented:
Closing this PR in preference to a simplified fix outlined in [PR 10579](https://github.com/openshift/console/pull/10579)
